gazebo_plugins/test/test_gazebo_ros_wheelslip.cpp

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gazebo/test/ServerFixture.hh>
+#include <gazebo_ros/node.hpp>
+#include <gazebo_ros/testing_utils.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+using namespace std::chrono_literals;
+
+/// Tests the gazebo_ros_wheelslip plugin
+class GazeboRosWheelSlipTest : public gazebo::ServerFixture
+{
+public:
+  void TestParameters(
+    std::string world_file,
+    std::string node_name,
+    std::map<std::string, double> parameter_pairs)
+  {
+    // Load the world
+    this->Load(world_file, true);
+    auto world = gazebo::physics::get_world();
+    ASSERT_NE(nullptr, world);
+    auto node = std::make_shared<rclcpp::Node>("test_gazebo_ros_wheelslip");
+    ASSERT_NE(nullptr, node);
+
+    // Create parameters client
+    auto parameters_client = std::make_shared<rclcpp::SyncParametersClient>(
+      node, node_name);
+
+    // Wait for serive to show up
+    int counter = 0;
+    while (!parameters_client->wait_for_service(1s)) {
+      std::cout << "service not available, waiting again..." << std::endl;
+      counter++;
+      if (counter > 5) {FAIL();}
+    }
+
+    // Verify the parameters
+    std::vector<std::string> parameter_names;
+    for (auto & element : parameter_pairs) {
+      parameter_names.push_back(element.first);
+    }
+    std::this_thread::sleep_for(1s);
```

Comment:
I don't think this sleep should be necessary. We're already waiting up to 5 seconds above for the service to be ready.

Reply:
Some of the tests are failing randomly if I remove this sleep (expected parameter is say 10.0, but we get 0.0 from the ``SyncParametersClient``). I think this is because the topic service becomes available after we declare the ``gazebo_ros::Node`` with the ``sdf``, but the parameters are set later in the plugin. If we query the parameters between these events, they will be returned as 0.

Reply:
Ah, that's tricky. I'm worried that this arbitrary sleep could be flaky or introduce extra time unnecessarily to the tests. IMO, it would be better to wait until the parameters match the expected values or timeout (e.g. continually call get_parameters until our condition is satisfied or a timeout occurs).

Reply:
Yeah that makes sense. I'll modify the tests accordingly

Reply:
Done